### PR TITLE
feat(chat) add message reactions

### PR DESCRIPTION
**Pull request by he-patrick**
- add chat buttons to allow message reactions and other functionalities
- render received reactions

### CODE REVIEW

File: react/features/chat/components/web/EmojiSelector.tsx [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import React from 'react';
+
+const EmojiSelector = ({ onSelect }) => {
+    const emojiMap = {
```
Comment (mihhu) [inline react/features/chat/components/web/EmojiSelector.tsx:4]: Until we make these options configurable, let's put here the commonly used reactions: 👍, ❤️,  😂, 😮 / 🔥, 😢, 🙏  or similar.

File: react/features/chat/types.ts [hunk 2]
```diff
@@ -44,6 +45,11 @@ export interface IChatMessageProps extends WithTranslation {
      */
     canReply?: boolean;
 
+    /**
+     * Whether the chat message menu is visible or not.
+     */
+    chatMessageMenu?: boolean;
```
Comment (mihhu) [inline react/features/chat/types.ts:51]: Is this used anywhere? I remember it was renamed from kebabMenu but I think we no longer need it

File: react/features/chat/components/web/ReactButton.tsx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        setAnchorEl(null);
+    };
+
+    const open = Boolean(anchorEl);
+    const id = open ? 'react-popover' : undefined;
+
+    return (
+        <div className = { classes.reactionPanelContainer }>
+            <Button
+                accessibilityLabel = { t('toolbar.accessibilityLabel.react') }
```
Comment (mihhu) [inline react/features/chat/components/web/ReactButton.tsx:63]: Please add this to the translation json file https://github.com/he-patrick/jitsi-meet/blob/master/lang/main.json#L1199

**Comment by mihhu** [comment]
There are a couple of lint and tsc errors left, please address them too 🙏 
https://github.com/jitsi/jitsi-meet/actions/runs/10473283350/job/29004816076?pr=15006

File: react/features/chat/types.ts (hunk 2 above)
Comment (he-patrick) [reply react/features/chat/types.ts:51]: It's needed to determine wether to display the kebab menu only (on the left side, for one's own message) or both kebab menu and reactions menu for other participant's messages.

File: react/features/chat/components/web/ChatMessage.tsx [hunk 4]
```diff
@@ -167,42 +255,151 @@ const ChatMessage = ({
...
+                    className = { classes.reactionBox }
+                    onClick = { handleReactionsClick }>
+                    {reactionsArray.slice(0, 3).map(({ reaction }, index) => (
+                        <span key = { index }>
+                            {reaction}
+                        </span>
+                    ))}
+                    {reactionsArray.length > 3 && (
+                        <span className = { classes.reactionCount }>
+                            +{reactionsArray.length - 3}
```
Comment (mihhu) [inline react/features/chat/components/web/ChatMessage.tsx:290]: I think we should count the number of reactions, not the number of participants that reacted with various reactions, wdys?

**Review by mihhu — COMMENTED** [review]
Aaaalmost done, I left some comments 🙌 

There's this thing I noticed when you were doing the demo too, did you have the chance to investigate it?
Pasted Graphic 1

File: react/features/chat/components/web/ChatMessage.tsx [hunk 5]
```diff
@@ -167,42 +255,151 @@ const ChatMessage = ({
...
+
+        const openReactions = Boolean(reactionsAnchorEl);
+        const reactionsId = openReactions ? 'reactions-popover' : undefined;
+
+        return (
+            <>
+                <div
+                    className = { classes.reactionBox }
+                    onClick = { handleReactionsClick }>
+                    {reactionsArray.slice(0, 3).map(({ reaction }, index) => (
```
Comment (saghul) [inline react/features/chat/components/web/ChatMessage.tsx:283]: Let's put this 3 in a constant please.

File: react/features/chat/components/web/ChatMessage.tsx (hunk 4 above)
Comment (saghul) [reply react/features/chat/components/web/ChatMessage.tsx:290]: WhatsApp counts the total reactions, FWIW.

File: react/features/chat/components/web/PrivateMessageButton.tsx [hunk 6]
```diff
@@ -22,7 +22,7 @@ export interface IProps {
     /**
      * The ID of the participant that the message is to be sent.
      */
-    participantID: string;
+    participantId: string;
```
Comment (saghul) [inline react/features/chat/components/web/PrivateMessageButton.tsx:25]: Please undo all changes to this file, they are unnecessary.

File: react/features/chat/components/web/KebabMenu.tsx [hunk 7]
```diff
@@ -0,0 +1,196 @@
+import Popover from '@mui/material/Popover';
```
Comment (saghul) [inline react/features/chat/components/web/KebabMenu.tsx:1]: Use our popover component.

File: react/features/chat/components/web/KebabMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+        if (isLobbyMessage) {
+            dispatch(handleLobbyChatInitialized(participantId));
+        } else {
+            dispatch(openChat(participant));
+        }
+        handleClose();
+    }, [ dispatch, isLobbyMessage, participant, participantId ]);
+
+    const handleCopyClick = useCallback(() => {
+        if (navigator.clipboard) {
```
Comment (saghul) [inline react/features/chat/components/web/KebabMenu.tsx:118]: Use our `copyText` helper function.

**Review by saghul — CHANGES_REQUESTED** [review]

File: react/features/chat/components/web/ChatMessage.tsx [hunk 9]
```diff
@@ -1,28 +1,43 @@
...
+        chatMessageFooterLeft: {
+            display: 'flex',
+            flexGrow: 1,
+            overflow: 'hidden'
+        },
         chatMessageWrapper: {
             maxWidth: '100%'
         },
-
         chatMessage: {
```
[the remark this answers is not in the document]
Comment (he-patrick) [reply react/features/chat/components/web/ChatMessage.tsx:41]: I tried setting the maxWidth, which works fine for the first message, but it causes this strange behaviour with following messages.

image

[the remark this answers is not in the document]
Comment (he-patrick) [reply react/features/chat/components/web/ChatMessage.tsx:41]: Sorry this picture is clearer: you can see that it prevents the following messages from filling up the space for some reason.

image

File: react/features/chat/components/web/ReactButton.tsx [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    }, [ onSendReaction, handleClose ]);
+
+    const popoverContent = (
+        <div className = { classes.popoverContent }>
+            <EmojiSelector onSelect = { handleEmojiSelect } />
+        </div>
+    );
+
+    return (
+        <Popover
```
Comment (mihhu) [inline react/features/chat/components/web/ReactButton.tsx:69]: I noticed the Popover goes out of the window if one of the participant names is too long:
image

Please try adding `maxWidth: '150px'` to reactionsPopover and `maxWidth: '120px'` to participantList, it should look a bit better:
image

File: react/features/chat/components/web/ChatMessage.tsx [hunk 11]
```diff
@@ -167,42 +261,144 @@ const ChatMessage = ({
...
+                    )}
+                </div>
+            </Popover>
+        );
+    };
+
     return (
         <div
-            className = { cx(classes.chatMessageWrapper, type) }
+            className = { cx(showDisplayName && shouldDisplayChatMessageMenu ? classes.chatFirstMessageWrapper : classes.chatMessageWrapper, type) }
```
[the remark this answers is not in the document]
Comment (mihhu) [reply react/features/chat/components/web/ChatMessage.tsx:330]: Please disregard^

I looked a bit more into the issue with the squished first message and the root cause is the participant name - when it's a lengthy one - it looks alright, name gets ellipsis and the message occupies 100% of the expected space, however when it's a short one, we're in trouble. So I say let's add `width: '130px'` to the displayName class and leave maxWidth: 100% for all chatMessageWrappers (an remove the ternary please).

[the remark this answers is not in the document]
Comment (he-patrick) [reply react/features/chat/components/web/ChatMessage.tsx]: I tried width: '130px' but I noticed that it made messages that are shorter set to that width too. maxWidth: '130px' seems to work well though.

**Review by mihhu — APPROVED** [review]

**Comment by mihhu** [comment]
🔥

**Comment by saghul** [comment]
@he-patrick Congrats Patrick, it's in! 🙌

**Pull request merged**